### PR TITLE
make total weight optional

### PR DESCRIPTION
**Pull request by zirain**
Signed-off-by: hejianpeng <hejianpeng2@huawei.com>

Commit Message: make WeightedCluster.TotalWeight optional
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] #21845
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by wrowe** [comment]
/wait

on release notes requested

File: changelogs/current.yaml [hunk 1]
```diff
@@ -42,6 +42,9 @@ behavior_changes:
     If there is no entry referred by the filter config name, the canonical filter name
     (e.g., *envoy.filters.http.buffer* for the HTTP buffer filter) will be used for the backwards
     compatibility.
+- area: router
+  change: |
+    :ref:`total_weight <envoy_v3_api_field_config.route.v3.WeightedCluster.total_weight>` could be optional.
```
Comment (mattklein123) [inline changelogs/current.yaml:47]:
```suggestion
    Weighted cluster's :ref:`total_weight <envoy_v3_api_field_config.route.v3.WeightedCluster.total_weight>` is now optional. If not set, Envoy will no longer validate that all weights add up to 100.
```

Also, you changed the meaning if 2 fields. Reference both above with similar language. Thank you!

/wait

Comment (zirain) [reply changelogs/current.yaml]: always got error:
```
generated/rst/version_history/v1.23/v1.23.0.rst:37: WARNING: undefined label: envoy_api_field_route.weightedcluster.total_weight
```

how can I fix this?

Comment (zirain) [reply changelogs/current.yaml]: I'm trying to add ref to this https://github.com/envoyproxy/envoy/pull/21892#discussion_r915983739, which seems like a v2 api?

Comment (mattklein123) [reply changelogs/current.yaml]: You can ignore v2, but you changed the documentation/defaults for 2 v3 fields, both of those should be linked to.

File: api/envoy/config/route/v3/route_components.proto [hunk 2]
```diff
@@ -381,7 +381,7 @@ message WeightedCluster {
     // An integer between 0 and :ref:`total_weight
     // <envoy_v3_api_field_config.route.v3.WeightedCluster.total_weight>`. When a request matches the route,
     // the choice of an upstream cluster is determined by its weight. The sum of weights across all
-    // entries in the clusters array must add up to the total_weight, which defaults to 100.
+    // entries in the clusters array must add up to the total_weight, if total_weight is greater than 0.
```
[the remark this answers is not in the document]
Comment (zirain) [reply api/envoy/config/route/v3/route_components.proto:392]: I'm a little confused, changes here is refer to `total_weight`, the behavior of `weight` is not changed.

[the remark this answers is not in the document]
Comment (mattklein123) [reply api/envoy/config/route/v3/route_components.proto:392]: What I mean is please document the semantic change to this field.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Comment by zirain** [comment]
CI is blocked?

image

**Pull request merged**